spatialos-sdk/src/worker/schema/collections.rs

```diff
@@ -0,0 +1,156 @@
+use crate::worker::schema::{Field, FieldId, SchemaObject};
+use spatialos_sdk_sys::worker::*;
+use std::{collections::BTreeMap, marker::PhantomData};
+
+/// Marker type corresponding to the [`option`] schemalang collection type.
+///
+/// `option<T>` is represented as [`Option<U>`][option] in the generated Rust code,
+/// where `U` is the Rust type corresponding to `T`. This type is named `Optional`
+/// instead of `Option` to avoid conflicting with Rust's built-in [`Option`][option]
+/// type.
+///
+/// See the [module-level documentation](index.html) for more information.
+///
+/// [`option`]: https://docs.improbable.io/reference/14.2/shared/schema/reference#collection-types
+/// [option]: https://doc.rust-lang.org/std/option/index.html
+pub struct Optional<T>(PhantomData<T>);
+
+impl<T> Field for Optional<T>
+where
+    T: Field,
+{
+    type RustType = Option<T::RustType>;
+
+    fn get_or_default(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        if T::count(object, field) > 0 {
+            Some(object.get::<T>(field))
+        } else {
+            None
+        }
+    }
+
+    fn add(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        if let Some(value) = value {
+            object.add::<T>(field, value);
+        }
+    }
+
+    fn index(_object: &SchemaObject, _field: FieldId, _index: usize) -> Self::RustType {
+        panic!("Optional fields cannot be indexed into")
+    }
+
+    fn count(_object: &SchemaObject, _field: FieldId) -> usize {
+        panic!("Optional fields cannot be counted")
```

Comment:
Feels weird that we panic here since optional values do have a count (either 0 or 1). Is this to indicate that this function shouldn't ever be called during serialization? 

If so, maybe this should be `unreachable!()`? Although that also feels a little wonky

Reply:
Optional fields can't be counted since you can't have a `list<option<T>>` in the schema. This function will be removed entirely once I split off a separate `IndexedField` trait 🙂

Reply:
Ah sounds good! :)